discretisedfield/util/util.py

```diff
@@ -17,71 +18,89 @@ def array2tuple(array):
     return array.item() if array.size == 1 else tuple(array.tolist())
 
 
-def as_array(mesh, dim, val):
-    array = np.empty((*mesh.n, dim))
-    if isinstance(val, numbers.Real) and (dim == 1 or val == 0):
-        # The array for a scalar field with numbers.Real value or any
-        # field with zero value.
-        array.fill(val)
-    elif isinstance(val, complex) and (dim == 1 or val == 0):
-        array = np.empty((*mesh.n, dim), dtype='complex128')
-        array.fill(val)
-    elif isinstance(val, np.ndarray) and val.shape == array.shape:
-        array = val
-    elif isinstance(val, (tuple, list, np.ndarray)):
-        if len(val) != dim:
-            msg = (f'Wrong dimension {len(val)} provided for value;'
-                   f' expected dimension is {dim}')
-            raise ValueError(msg)
-        if (isinstance(val, np.ndarray) and val.dtype == 'complex128'
-                or np.iscomplex(val).any()):
-            array = np.empty((*mesh.n, dim), dtype='complex128')
+def as_array(mesh, dim, val, dtype=np.float64):
+    array = np.empty((*mesh.n, dim), dtype=dtype)
+    return _fill_array(val, array, mesh, dim)
+
+
+@functools.singledispatch
+def _fill_array(val, array, mesh, dim):
+    raise ValueError('Unsupported type {type(val)}.')
+
+
+@_fill_array.register(numbers.Complex)  # contains numbers.Real
+def _(val, array, mesh, dim):
+    # TODO why not allow uniform values for vector fields
+    if dim > 1 and val != 0:
+        raise ValueError('Wrong dimension 1 provided for value;'
+                         f' expected dimension is {dim}')
+    array.fill(val)
+    return array
+
+
+@_fill_array.register(list)
+@_fill_array.register(tuple)
+def _(val, array, mesh, dim):
+    if len(val) != dim:
+        raise ValueError(f'Wrong dimension {len(val)} provided for value;'
+                         f' expected dimension is {dim}')
+    array[..., :] = val
```

Comment:
I guess that using `np.full` would be broadcasted correctly and this would probably reduce the number of separate implementations.